cli_tools/gce_vm_image_import/importer/importer.go

```diff
@@ -85,27 +95,45 @@ func (i *importer) runInflate(ctx context.Context) (err error) {
 	case <-ctx.Done():
 		err = ctx.Err()
 	default:
+		finished := i.handleCancel(ctx, i.inflater.cancel)
 		i.pd, err = i.inflater.inflate(ctx)
 		i.traceLogs = append(i.traceLogs, i.inflater.traceLogs()...)
+		close(finished)
 	}
 	return err
 }
 
-func (i *importer) runProcess(ctx context.Context) (err error) {
-	processor, err := i.processorProvider.provide(i.pd)
+func (i *importer) runProcess(ctx context.Context) error {
+	var err error
+	i.processor, err = i.processorProvider.provide(i.pd)
 	if err != nil {
 		return err
 	}
 	select {
 	case <-ctx.Done():
 		err = ctx.Err()
 	default:
-		err = processor.process(ctx)
-		i.traceLogs = append(i.traceLogs, processor.traceLogs()...)
+		finished := i.handleCancel(ctx, i.processor.cancel)
+		err = i.processor.process(ctx)
+		i.traceLogs = append(i.traceLogs, i.processor.traceLogs()...)
+		close(finished)
 	}
 	return err
```

Comment:
My understanding of the context pattern is that it's the callee's responsibility to monitor the context, and to quit when the context is cancelled. The immediate result is that an implementation like the data disk processor will "just work" (assuming that the client that it's using also follows this pattern).

The other benefit is that we keep the `importer.go` file focused on business logic, both in its implementation, its future reading, and its  testers. This will become *more* important when we add upcoming features such as:
* UEFI
* native vmdk
* OS inspection
* capacity inspection

Reply:
Can you be more specific, in the context of this code?

Reply:
Assuming you initialize the context's timeout prior to executing `importer.Run`, you need zeror changes to `importer.go` or its test.

Reply:
There are a couple of issues with this approach:
- logging to GCS fails on context expiration leading to customer visible errors and a partial loss of logs
- I don't see an easy way to change Daisy step failure reason to indicate a timeout occurred, which is one of the main requirements for this change as it allows us to track timeouts vs. other failures.
- temporary disk cleanup won't happen because importer.diskCleanup() is initialized only after inflation finishes. Disk is created during inflation. If timeout happens during inflation, disk will be left hanging in the project. This might be solvable by initializing diskCleanup early and checking if the disk still exists or not.

Reply:
1. I'm not sure a single, global context would be appropriate for all RPCs. I imagine one would be used for the workflow, and would have the global timeout, and another would be used for individual operations such as logging. For example, you might want to limit logging to `n` amount of time. The [Polly framework](https://github.com/App-vNext/Polly/wiki/Transient-fault-handling-and-proactive-resilience-engineering) has a good discussion about various resiliency strategies, and does a good job of showing that we need multiple strategies since individual problems are slightly different.
2. For reporting the cancellation reason, doesn't the `context` package have typed errors for determining why the context was cancelled? See the [variables section](https://golang.org/pkg/context/#pkg-variables) in their docs. Also, we're free to implement the context.Context interface if those typed errors aren't sufficient.
3. Regarding cleanup of temporary disks, I like the solution that you've developed here; we can use that, right?
  * Install the disk cleanup deferal prior to starting inflation
  * If disk cleanup returns an HTTP 404, consider cleanup successful.

I have a proposal though! If we update Daisy so that cancellation [the context passed to Run](https://github.com/GoogleCloudPlatform/compute-image-tools/blob/423907b63e7430a8e9dbe5edb296e4811f73797a/daisy/workflow.go#L266) is well-behaved, then your global timeout feature is nearly free.

I foresee the following changes:
* Daisy:
  1. Ensure the correct cancellation reason is surfaced.
  2. Ensure that cleanup occurs after cancellation.
* importer package:
  1. Tune disk cleanup to ensure disks are removed if timeout occurs during inflation.
  2. Ensure that logging occurs regardless of cancellation.

I'm happy to divide up work to make it go quicker.

Also, I want to say that I appreciate the amount of time you've spent thinking about this problem, and how you've developed several approaches and solutions. :)

Reply:
1. What is the benefit of using separate context objects for each operation? Seems convoluted and not really plug&play as you describe it. We'd have a bunch of context objects, with different expiration times. It just seems like a future mess waiting to happen.
2. Contex has only cancelled/deadline variables. Also, we don't use context cancellation for user cancellation (we use w.Cancel channel). So that would need to be refactored as well.

I don't see how modifying Daisy to support context timeout is free: we are just shifting work from wrappers to Daisy with the amount of effort being currently unknown. We would be investing probably a week of extra work and possibly introducing bugs for benefits that are not really clear. 

What is your main concern with this PR? That timeout is not "free"?

Reply:
> What is your main concern with this PR? That timeout is not "free"?

As a new teammate, I was intimidated by the [previous iteration](https://github.com/GoogleCloudPlatform/compute-image-tools/blob/v20191206/cli_tools/gce_vm_image_import/importer/importer.go#L243), and my goal was to help future readers. The tactic was twofold. First streamline the program's entry, keeping it focused on business steps, while keeping mechanics (logging, timeout, retry) orthogonal.  Second, use idiomatic Go where possible.

That said, I've inferred that you are happy with the approach in this PR. To move us forward, I've given a few concrete review comments. Once those are done, I'd say it LGTM :)